numpy/ma/core.py

```diff
@@ -5079,24 +5007,66 @@ def mean(self, axis=None, dtype=None, out=None):
         1.5
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = self._data.mean(axis, dtype, out, **kwargs)
+
+            to_mask = ~np.isfinite(ret)
+            if self.size == 0:
+                to_mask = True
+
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    out.__setmask__(to_mask if np.any(to_mask) else False)
+                return out
+
+            if isinstance(ret, ndarray):
+                ret = ret.view(type(self))
+                ret.__setmask__(to_mask if np.any(to_mask) else False)
+                return ret
+            else:
+                # otherwise it's a scalar
+                return ret if self.size else masked
+
+        # code largely copied from _mean in numpy/core/_methods.py
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_float_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
We did indeed discuss `subok`, but I'm not sure it was for this part...

In any case, without the `subok`, this will not work if the masked array holds a quantity. Since this currently does work, it would thus be somewhat of a regression (although not a particularly awful one, since these masked quantities work only very partially). So, my suggestion would be to just do the in-place division -- which has the advantages that (1) it keeps the code closer to what it is now, thus avoiding other potential regressions, and (2) that it leaves any special tricks that are needed for division to the subclasses itself.

As for what works and does not presently, see the following (note that `np.core._methods._mean` is not OK, but `np.mean` is OK with quantities, because it will call a subclass's `mean` method):
```
import numpy as np, astropy.units as u
np.__version__
# '1.10.4'
q = np.arange(1,101) * u.m
mq = np.ma.MaskedArray(q)
mq.mean()
# masked_Quantity(data = 50.5 m,
#                 mask = False,
#           fill_value = 1e+20)
np.mean(q)
# <Quantity 50.5 m>
np.core._methods._mean(q)
# array(50.5)
```
With your version, the masked array method stops working:
```
mq.mean()
# array(50.5)
```